### DISCUSSION TITLE
Almost there: Release 0.4.0 🚀

### DISCUSSION THREAD

**Discussion by bcakmakoglu** [bcakmakoglu/vue-flow#33] [Announcements]
Hello everyone,

It has been quite a while since I started working on release 0.4.0 and I am sorry if you had to wait for features or bugfixes that have been scheduled for 0.4.0.
But it is almost here, the docs are in the process of being rewritten but examples have been fully updated so anyone wanting to try it out can do so by using the next-tag (i.e. `npm i @braks/vue-flow@next`.

There are a number of changes, some of them breaking, but ultimately it should ease up the whole usage of Vue Flow and make it a lot nicer to use with the composition API, which is the recommended approach.

I want to use this post to provide you all with a list of changes that are to be expected when migrating from 0.3.0 to 0.4.0,
so features, fixes and other changes will be listed here for discussion.
I will update this post in the coming days/week and document everything I can.

## [Link to updated Docs](https://next.vueflow.dev/)

Thanks for reading and using Vue Flow!

**Comment by bcakmakoglu** [comment]
# New Features 🚀 

## Nested Nodes
The main new feature coming with 0.4.0 is nested nodes as discussed in #20.

![grafik](https://user-images.githubusercontent.com/78412429/158214404-23d6154c-aa02-4234-b84a-c3163ada9ba7.png)

The API description has changed, so I removed the example here.
A new one will replace it soon.

## useVueFlow

The next rather big change and also feature is that basically the whole API for Vue Flow is available through the `useVueFlow` composable.
Meaning, you can hook into any event, access any store value, find any utility you might look for (like adding nodes, adding edges) by using this composable.
This also makes it easy to access Vue Flow's internal properties from outside it's context.

```vue
<script lang="ts" setup>
import { VueFlow, MiniMap, Controls, Background, BackgroundVariant, isNode, useVueFlow, Elements } from '~/index'

const elements = ref<Elements>([
  { id: '1', type: 'input', label: 'Node 1', position: { x: 250, y: 5 } },
  { id: '2', label: 'Node 2', position: { x: 100, y: 100 } },
  { id: '3', label: 'Node 3', position: { x: 400, y: 100 } },
  { id: '4', label: 'Node 4', position: { x: 400, y: 200 } },
  { id: 'e1-2', source: '1', target: '2', animated: true },
  { id: 'e1-3', source: '1', target: '3' },
])

// creates a new vue flow instance 
const { onPaneReady, onNodeDragStop, onConnect, instance, addEdges } = useVueFlow()
onPaneReady(({ fitView }) => {
  fitView({ padding: 0.1 })
})
onNodeDragStop((e) => console.log('drag stop', e))
onConnect((params) => addEdges([params]))

const updatePos = () =>
  elements.value.forEach((el) => {
    if (isNode(el)) {
      el.position = {
        x: Math.random() * 400,
        y: Math.random() * 400,
      }
    }
  })

const logToObject = () => console.log(instance.value?.toObject())
const resetTransform = () => elements.value.push({ id: '1234', position: { x: 50, y: 50 }, label: 'Foobar' })
const toggleclass = () => elements.value.forEach((el) => (el.class = el.class === 'light' ? 'dark' : 'light'))
</script>
<template>
  <VueFlow
    v-model="elements"
    class="vue-flow-basic-example"
    :default-zoom="1.5"
    :min-zoom="0.2"
    :max-zoom="4"
    :zoom-on-scroll="false"
  >
  </VueFlow>
</template>
```

**Reply by bcakmakoglu** [reply]
[Check the changelog for all changes.](https://github.com/bcakmakoglu/vue-flow/blob/feat/docs/CHANGELOG.md)

**Comment by ulissepress** [comment]
Any plan when this 0.4 release will be published? Nested groups is something we are waiting for... :-)

**Reply by bcakmakoglu** [reply]
Yeah it will be very soon, I'm currently testing if everyone works as expected and I gotta update the docs gradually.
You can go ahead and try out the pre-release though (`npm i @braks/vue-flow@next`)
Look into the examples to get an idea on how to use it, esp. nested nodes as the API has changed.

**Comment by ulissepress** [comment]
thank you!
